### PR TITLE
C++: Add more barriers to `cpp/tainted-arithmetic`

### PR DESCRIPTION
**Pull request by MathiasVP**
This PR does two things:
- First, it converts `cpp/tainted-arithmetic` to a `path-problem` query. This is mostly to help investigate results on LGTM.  It seems like a good fit for this query anyway.
- Second, it puts the `bounded` predicate (which was introduced in https://github.com/github/codeql/pull/5887 for `cpp/uncontrolled-arithmetic`) in a separate QL file so that it can also be used by `cpp/tainted-arithmetic`.

We should also be able to use it in `cpp/uncontrolled-allocation-size` once https://github.com/github/codeql/pull/6120 is merged (which this PR currently conflicts with).

In addition, as this query is kinda noisy I've added another barrier to block flow out of "small integer" types.

Since it's difficult to do an LGTM difference run I've just manually compared the "before state" with the "after state".

Before: https://lgtm.com/query/519140448821761310/
After: https://lgtm.com/query/2136072651063415759/

The only real noticeable difference is that we remove all the FPs on `mldbai/giflib` and `adriweb/tivars_lib_cpp`. These results are in fact removed by the `e.getUnspecifiedType().(IntegralType).getSize() <= 1` barrier.

The two test failures are caused by the conversion to `path-problem` in the internal testcases. I've fixed them in a PR in the internal repo that should be merged with this PR.

### CODE REVIEW

**Comment by geoffw0** [comment]
LGTM, subject to resolving conflicts with https://github.com/github/codeql/pull/6120 after it is merged.

> The only real noticeable difference is that we remove all the FPs on mldbai/giflib and adriweb/tivars_lib_cpp. These results are in fact removed by the e.getUnspecifiedType().(IntegralType).getSize() <= 1 barrier.

With 1 byte types are people just being more careful and deliberate about overflow?

I'm surprised the cases in `mldbai/giflib` at least aren't caught by `bounded` and the bitwise `&` case anyway.

**Comment by MathiasVP** [comment]
> I'm surprised the cases in `mldbai/giflib` at least aren't caught by `bounded` and the bitwise `&` case anyway.

Good point! I took another look at the results, and it turns out that the `&` case indeed removes the results related to the `&` cases.
However, there are a couple of results in the same project that ends at the expression `Buf[0]--`, which are removed by the barrier that blocks flow through small integers. As these will underflow to at most a value of `255`, I think it's fine to remove these alerts.

**Comment by MathiasVP** [comment]
This PR doesn't need a change-note as the change-note added in https://github.com/github/codeql/pull/5667 also perfectly covers the changes made in PR. 🎉

**Comment by MathiasVP** [comment]
> LGTM, subject to resolving conflicts with #6120 after it is merged.
> 
> >

Conflicts have been resolved. I've also rebumped the internal PR's submodule.

**Comment by MathiasVP** [comment]
The tests on this PR, and the internal one, are green now @geoffw0.

File: cpp/ql/src/Security/CWE/CWE-190/Bounded.qll [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+pragma[inline]
+private predicate boundedBitwiseAnd(Expr e, Expr andExpr, Expr operand1, Expr operand2) {
+  operand1 != operand2 and
+  e = operand1 and
+  upperBound(operand2.getFullyConverted()) < exprMaxVal(andExpr.getFullyConverted())
+}
+
+/**
+ * Holds if `e` is an arithmetic expression that cannot overflow, or if `e` is an operand of an
+ * operation that may greatly reduces the range of possible values.
```
Comment (geoffw0) [inline cpp/ql/src/Security/CWE/CWE-190/Bounded.qll:43]:
```suggestion
 * operation that may greatly reduce the range of possible values.
```

**Review by geoffw0 — COMMENTED** [review]
LGTM.

I suspect we may be able to remove the exclusion for 1-byte types at some point, when we have other improvements in place, but if it's doing good now I don't oppose it.

File: cpp/ql/src/Security/CWE/CWE-190/Bounded.qll (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-190/Bounded.qll]: Thanks!

**Review by geoffw0 — APPROVED** [review]
Thanks, :+1:

**Pull request merged**